message/infrastructure/http/retry.go

```diff
@@ -0,0 +1,83 @@
+package http
+
+import (
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrNonPositiveNumberOfRetries  = errors.New("number of retries should be positive")
+	ErrNonPositiveTimeToFirstRetry = errors.New("time to first retry should be positive")
+)
+
+type RetryPublisherConfig struct {
+	MaxRetries int
+	// each subsequent retry doubles the time to next retry.
+	TimeToFirstRetry time.Duration
+}
+
+func (c *RetryPublisherConfig) setDefaults() {
+	if c.MaxRetries == 0 {
+		c.MaxRetries = 5
+	}
+
+	if c.TimeToFirstRetry == 0 {
+		c.TimeToFirstRetry = time.Second
+	}
+}
+
+func (c RetryPublisherConfig) validate() error {
+	if c.MaxRetries <= 0 {
+		return ErrNonPositiveNumberOfRetries
+	}
+	if c.TimeToFirstRetry <= 0 {
+		return ErrNonPositiveTimeToFirstRetry
+	}
+
+	return nil
+}
+
+type RetryPublisher struct {
+	pub    *Publisher
+	config RetryPublisherConfig
+}
+
+func NewRetryPublisher(pub *Publisher, config RetryPublisherConfig) (*RetryPublisher, error) {
+	config.setDefaults()
+
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid RetryPublisher config")
+	}
+
+	return &RetryPublisher{
+		pub,
+		config,
+	}, nil
+}
+
+func (p RetryPublisher) Publish(topic string, messages ...*message.Message) error {
```

Comment:
it is pretty simple to test (and to broke :D), so you know :poultry_leg:

Reply:
2250244364ee79e11de27e1cf0ada0db189fa4d4